src/cpp/configoptions.cpp

```diff
@@ -0,0 +1,216 @@
+}
+
+// Instantiate ConfigOptions, getting selections from the default source
+ConfigOptions* ConfigOptions::create_from_default()
+{
+    return new ConfigOptions(
+        _default_source, _default_cfg_string);
+}
+
+// Set environment variables with a particular prefix as the default source
+void ConfigOptions::set_default_from_environment(const std::string& db_prefix)
+{
+    _default_source = cs_envt;
+    _default_cfg_string = db_prefix;
+}
+
+// Set a UTF-8 file containing JSON data as the default configuration source
+void ConfigOptions::set_default_from_file(const std::string& filename)
+{
+    _default_source = cs_file;
+    _default_cfg_string = filename;
+}
+
+// Set a string containing a JSON blob as the default configuration source
+void ConfigOptions::set_default_from_string(const std::string& json_blob)
+{
+    _default_source = cs_blob;
+    _default_cfg_string = json_blob;
+}
+
+// Retrieve the value of a numeric configuration option
+int64_t ConfigOptions::get_integer_option(
+    const std::string& key, int64_t default_value)
+{
+    // If we already have the value, return it
+    auto search = _int_options.find(key);
+    if (search != _int_options.end())
+        return _int_options[key];
+
+    // If we're doing lazy evaluation of keys, fetch the value
+    int64_t result = default_value;
+    if (_lazy) {
+        int temp = 0;
+        get_config_integer(temp, key, default_value);
+        result = (int64_t)temp;
+    }
+
+    // Store the final value before we exit
+    _int_options.insert({key, result});
+    return result;
+}
+
+// Retrieve the value of a string configuration option
+std::string ConfigOptions::get_string_option(
+    const std::string& key, const std::string& default_value)
+{
+    // If we already have the value, return it
+    auto search = _string_options.find(key);
+    if (search != _string_options.end())
+        return _string_options[key];
+
+    // If we're doing lazy evaluation of keys, fetch the value
+    std::string result(default_value);
+    if (_lazy) {
+        get_config_string(result, key, default_value);
+    }
+
+    // Store the final value before we exit
+    _string_options.insert({key, result});
+    return result;
+}
+
+// Retrieve the value of a boolean configuration option
+bool ConfigOptions::get_boolean_option(
+    const std::string& key, bool default_value)
+{
+    // If we already have the value, return it
+    auto search = _bool_options.find(key);
+    if (search != _bool_options.end())
+        return _bool_options[key];
+
+    // If we're doing lazy evaluation of keys, fetch the value
+    bool result = default_value;
+    if (_lazy) {
+        get_config_bool(result, key, default_value);
+    }
+
+    // Store the final value before we exit
+    _bool_options.insert({key, result});
+    return result;
+}
+
+// Check whether a configuration option is set in the selected source
+bool ConfigOptions::is_defined(const std::string& key)
+{
```

Comment:
Since these values are all flat configuration values, could we avoid doing 3 storage locations and just keep them all as strings internally w/a cast on retrieval?

Reply:
We could, but that would be exceeding ugly